crates/utils/src/lib.rs

```diff
@@ -24,6 +25,17 @@ macro_rules! string_vec {
     }};
 }
 
+#[cached]
+pub fn get_workspace_root() -> PathBuf {
```

Comment:
This is super handy so we don't have to pass it around as an argument or struct field, which got very convoluted.